File: framework/languages/node/e2e/ResilienceLifecycle/Server/Provider/Configuration/server-options.ts
export interface ServerOptions {
  readonly role: string;
  readonly httpUrl: string;
  readonly logDir: string;
  readonly evidenceFile?: string;
  readonly rid: string;
  readonly registryRouterEndpoint?: string;
  readonly channelEndpoint?: string;
}

export function parseServerOptions(args: readonly string[], defaultRole = 'provider'): ServerOptions {
  const values = new Map<string, string>();
  for (let i = 0; i < args.length; i += 1) {
    const key = args[i];
    if (!key.startsWith('--')) {
      continue;
    }
    if (i + 1 >= args.length) {
      throw new Error(`Missing value for ${key}.`);
    }
    values.set(key.slice(2), args[++i]);
  }

  const rid = values.get('rid') ?? 'node';
  process.env.ZLINK_E2E_RID = rid;
  return {
    role: defaultRole,
    httpUrl: values.get('http-url') ?? 'http://127.0.0.1:0',
    logDir: values.get('log-dir') ?? '/tmp/zlink-node-e2e-log',
    evidenceFile: values.get('evidence-file'),
    rid,
    registryRouterEndpoint: values.get('registry-router-endpoint'),
    channelEndpoint: values.get('channel-endpoint')
  };
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Server/Provider/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/provider/Configuration/ServerOptions.kt
package systems.zlink.e2e.kotlin.registrymessaging.provider.Configuration

data class ServerOptions(
    val rid: String,
    val httpUrl: String,
    val logDir: String,
    val evidenceFile: String?,
    val registryRouterEndpoint: String,
    val channelEndpoint: String?,
    val manualClientEndpoint: String?,
    val routeEndpoint: String?,
    val weight: Int,
    val routePeers: List<String>,
) {
    companion object {
        fun parse(args: Array<String>): ServerOptions {
            val values = linkedMapOf<String, MutableList<String>>()
            var index = 0
            while (index < args.size) {
                val key = args[index]
                if (!key.startsWith("--")) {
                    index++
                    continue
                }
                require(index + 1 < args.size) { "Missing value for $key." }
                values.getOrPut(key.removePrefix("--")) { mutableListOf() }.add(args[index + 1])
                index += 2
            }
            val rid = values.last("rid", "node")
            System.setProperty("zlink.e2e.rid", rid)
            return ServerOptions(
                rid = rid,
                httpUrl = values.last("http-url", "http://127.0.0.1:0"),
                logDir = values.last("log-dir", System.getProperty("java.io.tmpdir") + "/zlink-kotlin-e2e-log"),
                evidenceFile = values.lastOrNull("evidence-file"),
                registryRouterEndpoint = values.lastRequired("registry-router-endpoint"),
                channelEndpoint = values.lastOrNull("channel-endpoint"),
                manualClientEndpoint = values.lastOrNull("manual-client-endpoint"),
                routeEndpoint = values.lastOrNull("route-endpoint"),
                weight = values.last("weight", "100").toInt(),
                routePeers = values["route-peer"]?.toList().orEmpty(),
            )
        }
    }
}

private fun Map<String, List<String>>.last(key: String, defaultValue: String): String =
    this[key]?.lastOrNull()?.takeIf { it.isNotBlank() } ?: defaultValue

private fun Map<String, List<String>>.lastOrNull(key: String): String? =
    this[key]?.lastOrNull()?.takeIf { it.isNotBlank() }

private fun Map<String, List<String>>.lastRequired(key: String): String =
    lastOrNull(key) ?: throw IllegalArgumentException("--$key is required.")
